### PR TITLE
Clean multisend_setup

### PR DESCRIPTION
**Pull request by alkino**

### CODE REVIEW

File: coreneuron/network/multisend_setup.cpp [hunk 1]
```diff
@@ -379,26 +378,23 @@ static void fill_multisend_lists(int use_phase2,
...
         for (auto& g: gid2in) {
             InputPreSyn* ps = g.second;
             int i = ps->multisend_phase2_index_;
-            if (i >= 0 && max_multisend_targets < targets_phase2[i]) {
-                max_multisend_targets = targets_phase2[i];
+            if (i >= 0) {
+                max_multisend_targets = std::max(max_multisend_targets, targets_phase2[i]);
             }
         }
+    }
```
Comment (alkino) [inline coreneuron/network/multisend_setup.cpp]: maybe we can delete the computing of those unused variables?
@nrnhines

File: coreneuron/network/multisend_setup.cpp [hunk 2]
```diff
@@ -118,39 +118,37 @@ static int* newintval(int val, int size) {
     return x;
 }
 
-static int* newoffset(int* acnt, int size) {
-    int* aoff = new int[size + 1];
+static std::vector<int> newoffset(const std::vector<int>& acnt) {
```
Comment (ohm314) [inline coreneuron/network/multisend_setup.cpp]: As you are going through the trouble of changing all these functions and their signatures, can you please start adding doxygen with a `\brief` and a description of the arguments and return?

File: coreneuron/network/multisend_setup.cpp [hunk 3]
```diff
@@ -118,39 +118,37 @@ static int* newintval(int val, int size) {
...
+    nrnmpi_int_alltoallv(scnt.data(), c.data(), rdispl.data(), rcnt.data(), c.data(), rdispl.data());
+    rdispl = newoffset(rcnt);
 }
 
 #define all2allv_perf 1
 // input s, scnt, sdispl ; output, newly allocated r, rcnt, rdispl
 static void
-all2allv_int(int* s, int* scnt, int* sdispl, int*& r, int*& rcnt, int*& rdispl, const char* dmes) {
+all2allv_int(const std::vector<int>& s, const std::vector<int>& scnt, const std::vector<int>& sdispl, std::vector<int>& r, std::vector<int>& rcnt, std::vector<int>& rdispl, const char* dmes) {
 #if all2allv_perf
```
Comment (ohm314) [inline coreneuron/network/multisend_setup.cpp]: looks like this is always defined... @pramodk any reason why this is done like this?

File: coreneuron/network/multisend_setup.cpp [hunk 4]
```diff
@@ -118,39 +118,37 @@ static int* newintval(int val, int size) {
...
     double tm = nrn_wtime();
 #endif
     int np = nrnmpi_numprocs;
     all2allv_helper(scnt, rcnt, rdispl);
-    r = newintval(0, rdispl[np]);
+    r = std::vector<int>(np, 0);
 
-    nrnmpi_int_alltoallv(s, scnt, sdispl, r, rcnt, rdispl);
+    nrnmpi_int_alltoallv(s.data(), scnt.data(), sdispl.data(), r.data(), rcnt.data(), rdispl.data());
```
Comment (ohm314) [inline coreneuron/network/multisend_setup.cpp]: Not for now, but to put into an issue, the wrappers can be a little more high-level and accept vectors and hide the data-unwrapping `.data()` inside.

File: coreneuron/network/multisend_setup.cpp (hunk 2 above)
Comment (alkino) [reply coreneuron/network/multisend_setup.cpp]: Of course ;)

File: coreneuron/network/multisend_setup.cpp (hunk 4 above)
Comment (alkino) [reply coreneuron/network/multisend_setup.cpp]: Not sure... If it's use by someone outside, it has to stay as C API.

Comment (ohm314) [reply coreneuron/network/multisend_setup.cpp]: good point, come to think of it - the modls could use this (and they are not c++ified yet). OK, anyway, for now, no change, but maybe still worth creating an issue.

File: coreneuron/network/multisend_setup.cpp (hunk 2 above)
Comment (alkino) [reply coreneuron/network/multisend_setup.cpp]: In fact there is already a lot of comments everywhere

**Comment by pramodk** [comment]
I will take a look at this today.

**Comment by ohm314** [comment]
All passing - up to @pramodk to double-check.

File: coreneuron/network/multisend_setup.cpp [hunk 5]
```diff
@@ -451,151 +406,130 @@ static int setup_target_lists(int use_phase2, int** r_return) {
...
-            tl->size += 1;
-        } else {
-            tl = new TarList();
-            tl->size = 1;
-            (*gid2tarlist)[r[i]] = tl;
+        // Now figure out the size of the target list for each distinct gid in r1.
+        for (const auto& gid: r1) {
+            auto tar = gid2tarlist[gid];
+            ++(tar->size);
         }
```
Comment (pramodk) [inline coreneuron/network/multisend_setup.cpp:436]: rest of the code is same but seems like some logic is changed for above section? could you elaborate?

**Comment by pramodk** [comment]
By the way, before merging this I would like to run https://github.com/nrnhines/tqperf (and possibly integrate that into our CI) to make sure we run heavy spike communication test before such change.

I am currently waiting for https://github.com/neuronsimulator/nrn/issues/475.

**Comment by pramodk** [comment]
Here is how we should test [nrnhines/tqperf](https://github.com/pramodk/tqperf) model:
* neuronsimulator/nrn#475 is merged now. Pull latest neutron master and install
* Run NEURON and it will generate spk000.dat
* Run CoreNEURON with multisend and it will generate out.dat
* Make sure spikes are same

i.e. This how I tased 

```bash
salloc -A proj16 -N 1 --ntasks-per-node=36 --exclusive  -p prod  -C "cpu|nvme" --time=8:00:00
module load unstable cmake bison flex python-dev hpe-mpi gcc/8.3.0 boost

git clone https://github.com/pramodk/tqperf
cd tqperf
mkdir mod && mv *.mod mod/

# compile mod files
$NRN-INSTALL/bin/nrnivmodl mod
$CORENRN-INSTALL/bin/nrnivmodl-core mod

# Run NEURON and then CoreNEURON
srun ./x86_64/special -c tstop=50 run.hoc -mpi
srun ./x86_64/special-core -e 50 -d coredat/ --mpi --multisend

# same spike with the master of coreneuron
diff -w out.dat spk000.dat
```

With this PR:

```
$ srun ./x86_64/special-core -e 50 -d coredat/ --mpi --multisend
 num_mpi=36
 num_omp_thread=1


 Duke, Yale, and the BlueBrain Project -- Copyright 1984-2019
 version id unimplemented

 Additional mechanisms from files
 invlfire.mod

 Memory (MBs) :             After mk_mech : Max 10.1719, Min 9.9961, Avg 10.0098
 num_mpi=36
 num_omp_thread=1

 Memory (MBs) :            After MPI_Init : Max 10.1719, Min 9.9961, Avg 10.0098
 Memory (MBs) :          Before nrn_setup : Max 10.2891, Min 10.0977, Avg 10.1425
MPT ERROR: Rank 26(g:26) received signal SIGSEGV(11).
	Process ID: 139436, Host: r1i6n0, Program: /gpfs/bbp.cscs.ch/data/project/proj16/kumbhar/pramod_scratch/tqperf/x86_64/special-core
	MPT Version: HPE MPT 2.21  08/29/19 00:34:00

MPT: --------stack traceback-------
MPT: Attaching to program: /proc/139436/exe, process 139436
MPT: (no debugging symbols found)...done.
MPT: 0x00002aaaab1bc179 in waitpid () from /lib64/libpthread.so.0
MPT: Missing separate debuginfos, use: debuginfo-install glibc-2.17-260.el7_6.6.x86_64 libbitmask-2.0-sgi720r52.rhel76.x86_64 libcpuset-1.0-sgi720r102.rhel76.x86_64 libibverbs-41mlnx1-OFED.4.6.0.4.1.46101.x86_64 libmlx4-41mlnx1-OFED.4.5.0.0.3.46101.x86_64 libmlx5-41mlnx1-OFED.4.6.0.0.4.46101.x86_64 libnl3-3.2.28-4.el7.x86_64 libnuma-3.0sgi-sgi716r61.rhel73.x86_64
MPT: (gdb) #0  0x00002aaaab1bc179 in waitpid () from /lib64/libpthread.so.0
MPT: #1  0x00002aaaab720776 in mpi_sgi_system (
MPT: #2  MPI_SGI_stacktraceback (
MPT:     header=header@entry=0x7fffffff7680 "MPT ERROR: Rank 26(g:26) received signal SIGSEGV(11).\n\tProcess ID: 139436, Host: r1i6n0, Program: /gpfs/bbp.cscs.ch/data/project/proj16/kumbhar/pramod_scratch/tqperf/x86_64/special-core\n\tMPT Version: "...) at sig.c:340
MPT: #3  0x00002aaaab720972 in first_arriver_handler (signo=signo@entry=11,
MPT:     stack_trace_sem=stack_trace_sem@entry=0x2aaab74c0080) at sig.c:489
MPT: #4  0x00002aaaab720d0b in slave_sig_handler (signo=11,
MPT:     siginfo=<optimized out>, extra=<optimized out>) at sig.c:565
MPT: #5  <signal handler called>
MPT: #6  0x00002aaaac60ee75 in __memcpy_ssse3_back () from /lib64/libc.so.6
MPT: #7  0x00002aaaab67d37b in MPI_SGI_alltoallv_sma (sendcnts=<optimized out>,
MPT:     comm=8, recvtype=3, rdispls=0x1d40c60, recvcnts=<optimized out>,
MPT:     recvbuf=0x1d40940, sendtype=3, sdispls=0x1d3c9c0, sendbuf=0x1d3cb00)
MPT:     at alltoallv.c:637
MPT: #8  MPI_SGI_alltoallv (sendbuf=0x1d3cb00, sendcnts=<optimized out>,
MPT:     sdispls=0x1d3c9c0, sendtype=3, recvbuf=0x1d40940,
MPT:     recvcnts=<optimized out>, rdispls=0x1d40c60, recvtype=3, comm=8)
MPT:     at alltoallv.c:770
MPT: #9  0x00002aaaab67e09e in PMPI_Alltoallv (sendbuf=0x1d3cb00,
MPT:     sendcnts=0x1d3c920, sdispls=<optimized out>, sendtype=sendtype@entry=3,
MPT:     recvbuf=recvbuf@entry=0x1d40940, recvcnts=recvcnts@entry=0x1d40a80,
MPT:     rdispls=0x1d40c60, recvtype=3, comm=8) at alltoallv.c:893
MPT: #10 0x00002aaaaaf68542 in coreneuron::nrnmpi_int_alltoallv (s=<optimized out>,
MPT:     scnt=<optimized out>, sdispl=<optimized out>, r=r@entry=0x1d40940,
MPT:     rcnt=rcnt@entry=0x1d40a80, rdispl=rdispl@entry=0x1d40c60)
MPT:     at /gpfs/bbp.cscs.ch/data/project/proj16/kumbhar/pramod_scratch/coreneuron/coreneuron/mpi/mpispike.cpp:281
MPT: #11 0x00002aaaaaf6a5f3 in coreneuron::all2allv_int(std::vector<int, std::allocator<int> > const&, std::vector<int, std::allocator<int> > const&, std::vector<int, std::allocator<int> > const&, char const*) ()
MPT:     at /gpfs/bbp.cscs.ch/data/project/proj16/kumbhar/pramod_scratch/coreneuron/coreneuron/network/multisend_setup.cpp:142
MPT: #12 0x00002aaaaaf6ab7b in coreneuron::setup_target_lists(bool) ()
MPT:     at /gpfs/bbp.cscs.ch/data/project/proj16/kumbhar/pramod_scratch/coreneuron/coreneuron/network/multisend_setup.cpp:429
MPT: #13 0x00002aaaaaf6b94b in coreneuron::nrn_multisend_setup_targets(bool, int*&, int*&) ()
MPT:     at /gpfs/bbp.cscs.ch/data/project/proj16/kumbhar/pramod_scratch/coreneuron/coreneuron/network/multisend_setup.cpp:260
MPT: #14 0x00002aaaaaf6a082 in coreneuron::nrn_multisend_setup() ()
MPT:     at /gpfs/bbp.cscs.ch/data/project/proj16/kumbhar/pramod_scratch/coreneuron/coreneuron/network/multisend.cpp:450
MPT: #15 0x00002aaaaaf50f12 in coreneuron::determine_inputpresyn() ()
MPT:     at /gpfs/bbp.cscs.ch/data/project/proj16/kumbhar/pramod_scratch/coreneuron/coreneuron/io/nrn_setup.cpp:631
MPT: #16 0x00002aaaaaf515b0 in coreneuron::nrn_setup(char const*, bool, bool, bool, char const*, char const*, double*) ()
MPT:     at /gpfs/bbp.cscs.ch/data/project/proj16/kumbhar/pramod_scratch/coreneuron/coreneuron/io/nrn_setup.cpp:770
MPT: #17 0x00002aaaaaf437f1 in coreneuron::nrn_init_and_load_data(int, char**, bool, bool, bool) ()
MPT:     at /gpfs/bbp.cscs.ch/data/project/proj16/kumbhar/pramod_scratch/coreneuron/coreneuron/apps/main1.cpp:271
MPT: #18 0x00002aaaaaf44080 in run_solve_core ()
MPT:     at /gpfs/bbp.cscs.ch/data/project/proj16/kumbhar/pramod_scratch/coreneuron/coreneuron/apps/main1.cpp:479
MPT: #19 0x00002aaaac4dd495 in __libc_start_main () from /lib64/libc.so.6
MPT: #20 0x000000000040077e in _start ()
MPT: (gdb) A debugging session is active.
```

File: coreneuron/network/multisend_setup.cpp (hunk 5 above)
Comment (alkino) [reply coreneuron/network/multisend_setup.cpp:436]: this is the place of the segfault, so let's remember what I did here

File: coreneuron/network/multisend_setup.cpp [hunk 6]
```diff
@@ -451,151 +406,134 @@ static int setup_target_lists(int use_phase2, int** r_return) {
...
     // list of target ranks for the gids it gets.
 
-    // scnt is number of input gids from target
-    scnt = newintval(0, nhost);
-    for (Gid2IPS::iterator gid2in_it = gid2in.begin(); gid2in_it != gid2in.end(); ++gid2in_it) {
-        int gid = gid2in_it->first;
-        ++scnt[gid % nhost];
-    }
+    {
+        // scnt is number of input gids from target
```
Comment (pramodk) [inline coreneuron/network/multisend_setup.cpp:410]:
```suggestion
        // scnt1 is number of input gids from target
```

File: coreneuron/network/multisend_setup.cpp [hunk 7]
```diff
@@ -451,151 +406,134 @@ static int setup_target_lists(int use_phase2, int** r_return) {
...
-    sdispl = newoffset(scnt, nhost);
-    s = newintval(0, sdispl[nhost]);
-    for (Gid2IPS::iterator gid2in_it = gid2in.begin(); gid2in_it != gid2in.end(); ++gid2in_it) {
-        int gid = gid2in_it->first;
-        s[sdispl[gid % nhost]++] = gid;
-    }
-    // Restore sdispl for the message.
-    del(sdispl);
-    sdispl = newoffset(scnt, nhost);
+        // s are the input gids from target to be sent to the various intermediates
```
Comment (pramodk) [inline coreneuron/network/multisend_setup.cpp:417]:
```suggestion
        // s1 are the input gids from target to be sent to the various intermediates
```

File: tests/jenkins/install_coreneuron.sh [hunk 8]
```diff
@@ -1,6 +1,7 @@
 #!/usr/bin/bash
 
 set -e
+set -x
```
[the remark this answers is not in the document]
Comment (alkino) [reply tests/jenkins/install_coreneuron.sh:4]: I want to let it. For debugging  all scripts should be -x. This it not verbose

**Comment by alkino** [comment]
hurray \o/

**Pull request merged**